server-utils/src/hosts.rs

```diff
@@ -2,10 +2,76 @@
 
 use std::ascii::AsciiExt;
 
-// TODO [ToDr] Pre-process hosts (so that they don't contain the path)
-
 /// Host type
-pub type Host = String;
+#[derive(Clone, Hash, PartialEq, Eq, Debug)]
+pub struct Host {
+	hostname: String,
+	port: Option<u16>,
+	as_string: String,
+}
+
+
+impl<T: AsRef<str>> From<T> for Host {
+	fn from(string: T) -> Self {
+		Host::parse(string.as_ref())
+	}
+}
+
+impl Host {
+	/// Creates a new `Host` given hostname and port number.
+	pub fn new(hostname: &str, port: Option<u16>) -> Self {
+		let hostname = Self::pre_process(hostname);
+		let string = Self::to_string(&hostname, port);
+
+		Host {
+			hostname: hostname,
+			port: port,
+			as_string: string,
+		}
+	}
+
+	/// Attempts to parse given string as a `Host`.
+	/// NOTE: This method always succeeds and falls back to sensible defaults.
+	pub fn parse(hostname: &str) -> Self {
+		let hostname = Self::pre_process(hostname);
+		let mut hostname = hostname.split(':');
+		let host = hostname.next().unwrap();
+		let port = hostname.next().and_then(|port| port.parse().ok());
+
+		Host::new(host, port)
+	}
+
+	fn pre_process(host: &str) -> String {
+		// Remove possible protocol definition
+		let mut it = host.split("://");
+		let protocol = it.next().unwrap();
```

Comment:
That line cannot fail. Any string will result in non-empty iterator when `split` is called. I will replace `unwraps` with `expect` though.